reword section on Nix store (#10)
nix-community/nix: doc/manual/src/design/store/store.md

# Nix Store
# Store

A Nix store is a collection of *store objects* referred to by *store paths*.
Every store also has a "store directory path", which is a path prefix used for various purposes.
A Nix store is a collection of [store objects](objects.md) and associated operations.

> There are many types of stores, but all of them at least respect this model.

Author: I would like to specify that at least a little bit, but would not even know a different type of store.

Reviewer: Start with
```
nix --extra-experimental-features nix-command describe-stores --json | jq keys
[
  "Dummy Store",
  "Http Binary Cache Store",
  "Legacy SSH Store",
  "Local Binary Cache Store",
  "Local Daemon Store",
  "S3 Binary Cache Store",
  "SSH Store"
]
```
Without `--json` it does markdown --- that markdown was intended to go in the manual along with some of the other auto-generated sections, but that has not happened yet.

Reviewer: Concretely, `cache.nixos.org` is a "Http binary cache store", and remote builders are "SSH Store"s or "Legacy SSH Store"s. None of these expose to the client (the client itself implements the store interface) the file system, and `cache.nixos.org` in using S3 in fact may not use it behind the scenes either.

Author: Thanks, that's very helpful.

> It defaults to `/nix/store`, but is in principle arbitrary.

Reviewer: As usual, I want to de-emphasize the filesystem, because even today not all stores use it!

The only time a file system *must* be involved with when performing builds, since builders are Unix processes. At all other times, the file system is non-normative and purely an implementation detail.

Author: Fully agree with deemphasizing the file system, and maybe I stuck too closely to your original.

I'm still struggling to find the right balance between being abstract and concrete, because ultimately we want to lead readers into dealing with Nix' concrete implementation and everyday things, which after all do happen in the file system.

The wording was aiming exactly there: The store objects themselves are just object, and _one way_ to access them is through the file system. The store directory drops out of that immediately, so I went from there to express it in terms of the file system projection.

Adding a hint to `/nix/store` seems appropriate in that context, because this is what people encounter very quickly when dealing with Nix, they see it a million times on `stdout`. How else would you approach this?

If we wanted to abstract that, we could just as well call the store directory something like "store identifier", because the whole point of tying objects to their store is to make sure that the objects' references actually exist. In any case I would still add the file system as an example to link it to everyday experience. Leaving it out makes it overly obscure, even if it may be the right level of abstraction in our context.

Edit: The other thing are store paths, where the term itself is tied to the file system. It's the only thing we have. Really it's just a store object identifier `(store, hash, name)` rendered as a path to provide access to the object's file system representation. I would not find reasonable to build an entire terminology for that higher layer of abstraction. Structurally, we already have that in the subsections on store objects and paths, and I think there is value to wording the introduction in a precise but non-academic style.

Author: > The only time a file system must be involved with when performing builds, since builders are Unix processes. At all other times, the file system is non-normative and purely an implementation detail.

Adding to the picture: The Unix process model is also what governs the executables we build. The moment we write store paths into executables, scripts, or configuration files, we have baked in the file system representation of the store for good. We cannot change those paths any more without breaking all kinds of things, like offsets and hashes, that the software managed by Nix may rely on internally.

In practice there is little alternative to interfacing through the file system short of building on a different operating system paradigm, and while I agree we should be clear this is not the only way in principle, it is the only way we currently deal with, however deliberately.

Reviewer: > I'm still struggling to find the right balance between being abstract and concrete, because ultimately we want to lead readers into dealing with Nix' concrete implementation and everyday things, which after all do happen in the file system.

But they don't all!

https://github.com/nix-community/nix/pull/10#discussion_r859393751 I think it would be good to discuss the concrete other stores first to anchor this conversation in the things a that actually exist, vs my hopes and dreams :)

> Adding a hint to /nix/store seems appropriate in that context, because this is what people encounter very quickly when dealing with Nix, they see it a million times on stdout. How else would you approach this?

Oh, I have no issue saying the default is `/nix/store`.

> Adding to the picture: The Unix process model is also what governs the executables we build. The moment we write store paths into executables, scripts, or configuration files, we have baked in the file system representation of the store for good. We cannot change those paths any more without breaking all kinds of things, like offsets and hashes, that the software managed by Nix may rely on internally.

This is true, but just need happen during building itself. When store objects are not involved in building, they can sit "out there in the ether". Indeed, the binary cache stores already act that way.

Similarly, the very notion of a store dir is kind of wrong, because there is no reason one cannot mix store objects for different store dirs in one binary cache.

All this would make more after the relocation section, where we go through the *existing* algorithm to substitute e.g. sources from `cache.nixos.org` to other store dirs, so their bootstrapping can work. See 3e3eaa90dd1aaf4684697f27726ec72aec75206c and 641c95070162595c71a1f775a2364cd8533e1c4b for some details.

The bottom line is the store dir is "really" a property of store objects that have references (either to themselves, or to other store objects). And with that is a coherence rule that all store objects in a closure must agree on a store dir. Store objects without references are trivially relocatable and so conceptually either don't have a store dir or have a wildcard store dir.

Since the implementation doesn't in fact have a notion of store object at all (only store paths, and side effects), I suppose it can't be said whether that part is in agreement or not with the implementation :). But when one thinks through the "operational semantics" of how CA derivations / `nix store make-content-addresed` and this cross store dir substitution work, I think this is basically the "denotation semantics" one is compelled to come up with.

I am very excited to fill out all the information to get that to that section, because it is the real highlight for me -- first present the rules for various things by the book, but with somewhat unclear motivation and then use all those "facts" to come up with the right conceptual model so that it all makes sense!

----

On the future side, https://github.com/NixOS/nix/issues/5729 https://github.com/NixOS/nix/issues/5025 may be of interest.

# Store

> A Nix store is a collection of [store objects](objects.md) and associated operations.

Reviewer: Technically this says a store "is a collection of...and...operations" but the operations are not collected I don't think?

How about?
```suggestion
A Nix store is a collection of [store objects](objects.md) that supports certain standard operations.
```

Author: If you say "certain", it has the opposite effect of being vague, as the word avoids saying explicitly what those are. If we say it explicitly afterwards, there is no point alluding to that these operations are specific.

I find "standard" problematic, because what does that standard refer to? Where is that standard defined? Since I know what you want to get at, I understand that you mean "these operations define the default store API", but that is not visible from the text. I would like to avoid wording it that way, because then we should just specify that API and not say "it has an API". Currently we're trying to give an informal overview, and give a prose specification in the other paragraph, and maybe at some point we should just add a class diagram for clarity?

How about that?
```suggestion
A Nix store is a collection of [store objects](objects.md) with associated operations.
```
It's semantically not exactly what you suggest, because now it sounds like the objects have operations. This is not wrong, therefore I'd be fine with it, but I see you wanted to say that the store (or the collection in its entirety) has operations.

Another variant, which I cannot formulate as a proper `suggestion` because a deleted line is in between:

```suggestion
A Nix store is a collection of [store objects](objects.md) that supports the following operations:

- It allows adding and retrieving store objects.
- It can perform builds, that is, transform build inputs using instructions from the build plans into build outputs.
- It keeps track of *references* between data and can therefore garbage-collect unused store objects.

Store objects can hold arbitrary data, and Nix makes no distinction if they are used as build inputs, build results, or build plans.
```

Less repetitive, and implies there are "operations" by just telling what they are:

```suggestion
A Nix store is a collection of [store objects](objects.md) which

- allows adding and retrieving store objects.
- can perform builds, that is, transform build inputs using instructions from the build plans into build outputs.
- keeps track of *references* between data and can therefore garbage-collect unused store objects.

Store objects can hold arbitrary data, and Nix makes no distinction if they are used as build inputs, build results, or build plans.
```

These store objects can hold arbitrary data, and Nix makes no distinction if they are used as build inputs, build results, or build plans.

> A Nix store allows to add and retrieve store objects.

Reviewer: Grammar nit. Perhaps
```suggestion
A Nix store allows one to store and retrieve store objects.
```
or
```suggestion
A Nix store allows storing and retrieving store objects.
```
?

Author: Thanks for the hint, TIL [grammatical structure of "allow"](https://english.stackexchange.com/a/60285).

I am very sensible to using the same word root too much within a given context. "Storing store objects in a store" makes my toe nails curl up. What is wrong with "add"?

```suggestion
A Nix store allows adding and retrieving store objects.
```
